c7n/resources/iam.py

```diff
@@ -840,10 +843,18 @@ def process(self, resources):
             try:
                 client.delete_role(RoleName=r['RoleName'])
             except client.exceptions.DeleteConflictException as e:
-                self.log.warning(
-                    "Role:%s cannot be deleted, must remove role from instance profile first"
-                    % r['Arn'])
-                error = e
+                if self.data.get('force', False):
+                    attached_policy = client.list_attached_role_policies(RoleName=r['RoleName'])
```

Comment:
this functionality should be composed from a set-policies action on a role, rather than inlining it into an exception handler.

Reply:
Don't exactly understand what you meant here. But I have moved the operation out of the exception handler

Reply:
that it should be a separate action to set the policies of a role, and that action should be reused here.

Reply:
Thanks! Looks like there's an `action: set-policy` already present for iam-role. But it requires the policy arn from the user and then sets/unsets based on the `state` flag.
In my case, we wouldn't exactly know what policies are attached to the role. Do you think it'll be better to modify the set-policy action such that it would detach all policies from the role?
Sample policy would look like this:
```
          actions:
            - type: set-policy
              state: detached
```
set-policy: https://github.com/cloud-custodian/cloud-custodian/blob/master/c7n/resources/iam.py#L761

Reply:
yes parameterizing set-policy so that it can be reused is a good path forward, it makes that action more useful, and reduces code duplication for role delete force.

Reply:
Got it! Thanks

Reply:
in this case have the set-policy action take a `*` should encapsulate best for policy disassociation to role